L79-cpp-async-response-reader-destruction.md

```diff
@@ -0,0 +1,123 @@
+has a virtual destructor, `std::default_delete` is specialized such that the
+destructor never actually runs. So the only subclasses that can actually be
+implemeneted without hacks are those like `grpc::ClientAsyncResponseReader`,
+which knows it is always allocated on an arena and doesn't care about its
+destructor running. This proposal provides a plan for fixing this so that other
+subclasses can count on their destructor running and their backing memory being
+freed, as with any other C++ class.
+
+## Background
+
+We currently [specialize](https://github.com/grpc/grpc/blob/7e14d23ab46e1b0924d6c3e40797ed1d587aee7a/include/grpcpp/impl/codegen/async_unary_call.h#L404-L408)
+`std::default_delete` for
+`ClientAsyncResponseReaderInterface` to do nothing. This means that, despite the
+fact that async stub methods return
+`std::unique_ptr<ClientAsyncResponseReaderInterface>`, it's impossible to have a
+subclass of that interface that is allocated on the heap or whose destructor has
+side effects without a hack that manages ownership _outside of_ the `unique_ptr`
+returned by the stub.
+
+This was introduced by
+[`dd36b153`](https://github.com/grpc/grpc/commit/dd36b15315cd691e86a94d4574bd9f3e3a33633f),
+which introduced the specialization for `grpc::ClientAsyncResponseReader`
+itself, and
+[`e8a61d63`](https://github.com/grpc/grpc/commit/e8a61d63b5a6db0f81b688481a9485d412e5a41e),
+which did so for the interface. These commits didn't document their intent, but
+I suspect it was a combination of the following facts:
+
+*   In production uses, `ClientAsyncResponseReaderHelper::Create`
+    [allocates](https://github.com/grpc/grpc/blob/7e14d23ab46e1b0924d6c3e40797ed1d587aee7a/include/grpcpp/impl/codegen/async_unary_call.h#L96-L99)
+    the `ClientAsyncResponseReader` on an arena.
+
+*   Although it's technically non-trivial, the `ClientAsyncResponseReader`
+    destructor doesn't actually have any important side effects.
+
+*   Therefore it's not necessary to run the destructor, and it's not correct to
+    use `delete` to free the backing memory.
+
+From this point of view, the original commit (`dd36b153`) is technically
+correct: `std::unique_ptr<ClientAsyncResponseReader>` doesn't actually need to
+do anything. But:
+
+*   Of course the objects may be deleted other ways. If I have a
+    `std::unique_ptr<ClientAsyncResponseReaderInterface>` (i.e. without a
+    specific deleter type), it would be natural for me to assume that I can do
+    `delete p.release()` to get the same effect as destroying the unique
+    pointer. But that is not true, and doing so will be undefined behavior.
+
+    So from this point of view `dd36b153` is not correct, or at least makes for
+    a surprising API.
+
+*   `ClientAsyncResponseReader` is not the only subclass of the interface. The
+    interface's destructor is virtual, so other subclasses may have destructors
+    with side effects. However they are never actually run! This is a huge
```

Comment:
> The interface's destructor is virtual, so other subclasses may have destructors
>  with side effects. However they are never actually run! ...

I would replace this with:

```
Even though subclasses can override the virtual destructor to add a side effect, their dtor does not actually run. This results in subtle bugs and confusing behavior, specially because   `delete p.release()` works differently than
 just letting a `std::unique_ptr p` go out of scope.
```

Reply:
Done.